# Домашнее задание к лекции "Статистика. Практика"

## Задание 1

Вернемся к [набору данных о видеоиграх](https://github.com/obulygin/pyda_homeworks/blob/master/stat_case_study/video_games_sales.csv).

Ответьте на следующие вопросы:

1) Как критики относятся к спортивным играм?  
2) Критикам нравятся больше игры на PC или на PS4?  
3) Критикам больше нравятся стрелялки или стратегии?  

Для каждого вопроса:
- сформулируйте нулевую и альтернативную гипотезы;
- выберите пороговый уровень статистической значимости;
- опишите полученные результаты статистического теста.

In [50]:
import pandas as pd
import numpy as np

In [51]:
games = pd.read_csv('video_games_sales.csv')
games

Rank                           Name                      basename  \
0          1                     Wii Sports                    wii-sports   
1          2              Super Mario Bros.              super-mario-bros   
2          3                 Mario Kart Wii                mario-kart-wii   
3          4  PlayerUnknown's Battlegrounds  playerunknowns-battlegrounds   
4          5              Wii Sports Resort             wii-sports-resort   
...      ...                            ...                           ...   
55787  55788                    Indivisible                   indivisible   
55788  55789                     Lost Ember                    lost-amber   
55789  55790                     Lost Ember                    lost-amber   
55790  55791                     Lost Ember                    lost-amber   
55791  55792                     Falcon Age                    falcon-age   

                  Genre ESRB_Rating Platform         Publisher  \
0                Sports           E      Wii          Nintendo   
1              Platform         NaN      NES          Nintendo   
2                Racing           E      Wii          Nintendo   
3               Shooter         NaN       PC  PUBG Corporation   
4                Sports           E      Wii          Nintendo   
...                 ...         ...      ...               ...   
55787      Role-Playing         NaN       PC         505 Games   
55788         Adventure          RP       PC   Mooneye Studios   
55789         Adventure          RP      PS4   Mooneye Studios   
55790         Adventure          RP     XOne   Mooneye Studios   
55791  Action-Adventure         NaN      PS4           Unknown   

              Developer  VGChartz_Score  Critic_Score  ...  NA_Sales  \
0          Nintendo EAD             NaN           7.7  ...       NaN   
1          Nintendo EAD             NaN          10.0  ...       NaN   
2          Nintendo EAD             NaN           8.2  ...       NaN   
3      PUBG Corporation             NaN           NaN  ...       NaN   
4          Nintendo EAD             NaN           8.0  ...       NaN   
...                 ...             ...           ...  ...       ...   
55787    Lab Zero Games             NaN           NaN  ...       NaN   
55788   Mooneye Studios             NaN           NaN  ...       NaN   
55789   Mooneye Studios             NaN           NaN  ...       NaN   
55790   Mooneye Studios             NaN           NaN  ...       NaN   
55791   Outerloop Games             NaN           NaN  ...       NaN   

       PAL_Sales  JP_Sales  Other_Sales    Year  Last_Update  \
0            NaN       NaN          NaN  2006.0          NaN   
1            NaN       NaN          NaN  1985.0          NaN   
2            NaN       NaN          NaN  2008.0  11th Apr 18   
3            NaN       NaN          NaN  2017.0  13th Nov 18   
4            NaN       NaN          NaN  2009.0          NaN   
...          ...       ...          ...     ...          ...   
55787        NaN       NaN          NaN  2019.0  11th Apr 19   
55788        NaN       NaN          NaN  2019.0  11th Apr 19   
55789        NaN       NaN          NaN  2019.0  11th Apr 19   
55790        NaN       NaN          NaN  2019.0  11th Apr 19   
55791        NaN       NaN          NaN     NaN  12th Apr 19   

                                                     url  status  \
0      http://www.vgchartz.com/game/2667/wii-sports/?...       1   
1      http://www.vgchartz.com/game/6455/super-mario-...       1   
2      http://www.vgchartz.com/game/6968/mario-kart-w...       1   
3      http://www.vgchartz.com/game/215988/playerunkn...       1   
4      http://www.vgchartz.com/game/24656/wii-sports-...       1   
...                                                  ...     ...   
55787  http://www.vgchartz.com/game/224682/indivisibl...       1   
55788  http://www.vgchartz.com/games/game.php?id=2246...       1   
55789  http://www.vgchartz.com/game/224690/l

## как критики относятся к спортивным играм?

H0: средняя оценка критиков для спортивных игр >=7, критикам нравятся спортивные игры
H1: средняя оценка критиков для спортивных игр <7, критикам не нравятся спортивные игры

alfa = 0.05
сравниваем значения выборки с константой - нужен одновыборочный тест
инересует сравнение в меньшую сторону - односторонний тест

In [52]:
from scipy import stats as st
alpha = 0.05
result = st.ttest_1samp(games[games.Genre == 'Sports']['Critic_Score'].dropna(), 7)
print(result)

if (result.statistic < 0) & (result.pvalue / 2 < alpha):
    print('Отвергаем нулевую гипотезу, среднее меньше 7')
else:
    print('Не отвергаем нулевую нулевую гипотезу, критикам нравятся спортивные игры')

Ttest_1sampResult(statistic=5.197156914436623, pvalue=2.671410902970997e-07)
Не отвергаем нулевую нулевую гипотезу, критикам нравятся спортивные игры


S > 0, p < 0.05, те с высокой долей вероятности мы можем утверждыть, что критики ставят оценку 7 и больше, те им нравятся спортивные игры

## Критикам нравятся больше игры на PC или на PS4?  

H0: средняя оценка критиков для игрв на PS4 >= средней оценки для игр на PC, PS4 нравится больше
H1: средняя оценка критиков для игрв на PS4 < средней оценки для игр на PC

alfa = 0.05
сравниваем среднее двух выборок
двухвыборочный односторонний тест

In [55]:
result = st.ttest_ind(games[games.Platform == 'PS4']['Critic_Score'].dropna(), games[games.Platform == 'PC']['Critic_Score'].dropna(), equal_var=False)
print(result)
alpha = 0.05
if (result.statistic <0) & (result.pvalue/2 < alpha):
    print('Отвергаем нулевую гипотезу')
else:
    print('Не отвергаем нулевую гипотезу, PS4 нравится больше')

Ttest_indResult(statistic=2.7394476056351627, pvalue=0.006931808250254211)
Не отвергаем нулевую гипотезу, PS4 нравится больше


S > 0, P < 0.05, c высокой долей вероятности можем утверждать, что вредняя оценка за PS4 выше PC, те PS4 нравится больше

## Критикам нравятся больше стрелялки или стратегии?  

H0: средняя оценка критиков для стрелялок >= средней оценки для стрвтегий, стрелялки нравится больше
H1: средняя оценка критиков для стрелялок < средней оценки для стратегий

alfa = 0.05
сравниваем среднее двух выборок
двухвыборочный односторонний тест

In [61]:
result = st.ttest_ind(games[games.Genre == 'Shooter']['Critic_Score'].dropna(), games[games.Genre == 'Strategy']['Critic_Score'].dropna(), equal_var=False)
print(result)
alpha = 0.05
print(result.pvalue/2)
if (result.statistic <0) & (result.pvalue/2 < alpha):
    print('Отвергаем нулевую гипотезу')
else:
    print('Не отвергаем нулевую гипотезу, стрелялки нравится больше')

Ttest_indResult(statistic=-1.6073949711166526, pvalue=0.10838786414223071)
0.05419393207111536
Не отвергаем нулевую гипотезу, стрелялки нравится больше


S < 0, P > 0.05, тут S < 0, те среднее по стрелялкам ниже среднего по стратегиям, но P Value чуть выше 0.05, поэтому мы не можем с высокой долей уверенности говорить, что это статистически значимо, поэтому не отвергаем нулевую гипотезу
Но если повысим альфа до 0.055, то нулевая гипотеза будет отвергнута

In [ ]:
games[games.Genre == 'Strategy']['Critic_Score'].mean()


In [ ]:
games[games.Genre == 'Shooter']['Critic_Score'].mean()

## Задание 2

Реализуйте базовую модель логистической регрессии для классификации текстовых сообщений (используемые данные [здесь](https://github.com/obulygin/pyda_homeworks/blob/master/stat_case_study/spam.csv)) по признаку спама. Для этого:

1) Привидите весь текст к нижнему регистру;  
2) Удалите мусорные символы;  
3) Удалите стоп-слова;  
4) Привидите все слова к нормальной форме;  
5) Преобразуйте все сообщения в вектора TF-IDF. Вам поможет следующий код:  

```
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf = TfidfVectorizer()
tfidf_matrix = tfidf.fit_transform(df.Message)
names = tfidf.get_feature_names()
tfidf_matrix = pd.DataFrame(tfidf_matrix.toarray(), columns=cname)
```

Можете поэкспериментировать с параметрами [TfidfVectorizer](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.TfidfVectorizer.html);  
6) Разделите данные на тестовые и тренировочные в соотношении 30/70, укажите `random_state=42`. Используйте [train_test_split](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html);  
7) Постройте модель [логистической регрессии](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html), укажите `random_state=42`, оцените ее точность на тестовых данных;  
8) Опишите результаты при помощи [confusion_matrix](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.confusion_matrix.html?highlight=confusion_matrix#sklearn.metrics.confusion_matrix);  
9) Постройте датафрейм, который будет содержать все исходные тексты сообщений, классифицированные неправильно (с указанием фактического и предсказанного).

In [1]:
import re
import pandas as pd

In [2]:
spam = pd.read_csv('Spam.csv')
spam

Category                                            Message
0         ham  Go until jurong point, crazy.. Available only ...
1         ham                      Ok lar... Joking wif u oni...
2        spam  Free entry in 2 a wkly comp to win FA Cup fina...
3         ham  U dun say so early hor... U c already then say...
4         ham  Nah I don't think he goes to usf, he lives aro...
...       ...                                                ...
5567     spam  This is the 2nd time we have tried 2 contact u...
5568      ham               Will ü b going to esplanade fr home?
5569      ham  Pity, * was in mood for that. So...any other s...
5570      ham  The guy did some bitching but I acted like i'd...
5571      ham                         Rofl. Its true to its name

[5572 rows x 2 columns]

In [3]:
spam_lower_case = [spam.lower() for spam in spam.Message]
spam_lower_case

['go until jurong point, crazy.. available only in bugis n great world la e buffet... cine there got amore wat...',
 'ok lar... joking wif u oni...',
 "free entry in 2 a wkly comp to win fa cup final tkts 21st may 2005. text fa to 87121 to receive entry question(std txt rate)t&c's apply 08452810075over18's",
 'u dun say so early hor... u c already then say...',
 "nah i don't think he goes to usf, he lives around here though",
 "freemsg hey there darling it's been 3 week's now and no word back! i'd like some fun you up for it still? tb ok! xxx std chgs to send, £1.50 to rcv",
 'even my brother is not like to speak with me. they treat me like aids patent.',
 "as per your request 'melle melle (oru minnaminunginte nurungu vettam)' has been set as your callertune for all callers. press *9 to copy your friends callertune",
 'winner!! as a valued network customer you have been selected to receivea £900 prize reward! to claim call 09061701461. claim code kl341. valid 12 hours only.',
 'had you

In [4]:
# уберем все служебные символы
only_words_text = [re.sub('[\W_]+',' ', text) for text in spam_lower_case]

only_words_text

['go until jurong point crazy available only in bugis n great world la e buffet cine there got amore wat ',
 'ok lar joking wif u oni ',
 'free entry in 2 a wkly comp to win fa cup final tkts 21st may 2005 text fa to 87121 to receive entry question std txt rate t c s apply 08452810075over18 s',
 'u dun say so early hor u c already then say ',
 'nah i don t think he goes to usf he lives around here though',
 'freemsg hey there darling it s been 3 week s now and no word back i d like some fun you up for it still tb ok xxx std chgs to send 1 50 to rcv',
 'even my brother is not like to speak with me they treat me like aids patent ',
 'as per your request melle melle oru minnaminunginte nurungu vettam has been set as your callertune for all callers press 9 to copy your friends callertune',
 'winner as a valued network customer you have been selected to receivea 900 prize reward to claim call 09061701461 claim code kl341 valid 12 hours only ',
 'had your mobile 11 months or more u r entitle

In [5]:
# разделим все на слова
texts_splitted = [text.split() for text in only_words_text]

texts_splitted

[['go',
  'until',
  'jurong',
  'point',
  'crazy',
  'available',
  'only',
  'in',
  'bugis',
  'n',
  'great',
  'world',
  'la',
  'e',
  'buffet',
  'cine',
  'there',
  'got',
  'amore',
  'wat'],
 ['ok', 'lar', 'joking', 'wif', 'u', 'oni'],
 ['free',
  'entry',
  'in',
  '2',
  'a',
  'wkly',
  'comp',
  'to',
  'win',
  'fa',
  'cup',
  'final',
  'tkts',
  '21st',
  'may',
  '2005',
  'text',
  'fa',
  'to',
  '87121',
  'to',
  'receive',
  'entry',
  'question',
  'std',
  'txt',
  'rate',
  't',
  'c',
  's',
  'apply',
  '08452810075over18',
  's'],
 ['u', 'dun', 'say', 'so', 'early', 'hor', 'u', 'c', 'already', 'then', 'say'],
 ['nah',
  'i',
  'don',
  't',
  'think',
  'he',
  'goes',
  'to',
  'usf',
  'he',
  'lives',
  'around',
  'here',
  'though'],
 ['freemsg',
  'hey',
  'there',
  'darling',
  'it',
  's',
  'been',
  '3',
  'week',
  's',
  'now',
  'and',
  'no',
  'word',
  'back',
  'i',
  'd',
  'like',
  'some',
  'fun',
  'you',
  'up',
  'for',
  'it',


In [6]:
import nltk
nltk.download('stopwords')
nltk.download('wordnet')

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\Maria\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\Maria\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


True

In [8]:
from nltk.corpus import stopwords
stopwords_set = set(stopwords.words('english'))
without_sw = [[word for word in text if word not in stopwords_set] for text in texts_splitted]
without_sw

[['go',
  'jurong',
  'point',
  'crazy',
  'available',
  'bugis',
  'n',
  'great',
  'world',
  'la',
  'e',
  'buffet',
  'cine',
  'got',
  'amore',
  'wat'],
 ['ok', 'lar', 'joking', 'wif', 'u', 'oni'],
 ['free',
  'entry',
  '2',
  'wkly',
  'comp',
  'win',
  'fa',
  'cup',
  'final',
  'tkts',
  '21st',
  'may',
  '2005',
  'text',
  'fa',
  '87121',
  'receive',
  'entry',
  'question',
  'std',
  'txt',
  'rate',
  'c',
  'apply',
  '08452810075over18'],
 ['u', 'dun', 'say', 'early', 'hor', 'u', 'c', 'already', 'say'],
 ['nah', 'think', 'goes', 'usf', 'lives', 'around', 'though'],
 ['freemsg',
  'hey',
  'darling',
  '3',
  'week',
  'word',
  'back',
  'like',
  'fun',
  'still',
  'tb',
  'ok',
  'xxx',
  'std',
  'chgs',
  'send',
  '1',
  '50',
  'rcv'],
 ['even', 'brother', 'like', 'speak', 'treat', 'like', 'aids', 'patent'],
 ['per',
  'request',
  'melle',
  'melle',
  'oru',
  'minnaminunginte',
  'nurungu',
  'vettam',
  'set',
  'callertune',
  'callers',
  'press'

In [9]:
from nltk.stem import WordNetLemmatizer
wordnet_lemmatizer = WordNetLemmatizer()
lemmatized_text = [[wordnet_lemmatizer.lemmatize(word) for word in text] for text in without_sw]
lemmatized_text[15]

['xxxmobilemovieclub',
 'use',
 'credit',
 'click',
 'wap',
 'link',
 'next',
 'txt',
 'message',
 'click',
 'http',
 'wap',
 'xxxmobilemovieclub',
 'com',
 'n',
 'qjkgighjjgcbl']

In [10]:
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf = TfidfVectorizer()
tfidf_matrix = tfidf.fit_transform(spam.Message)
names = tfidf.get_feature_names()
tfidf_matrix = pd.DataFrame(tfidf_matrix.toarray(), columns=names)
tfidf_matrix

00  000  000pes  008704050406  0089  0121  01223585236  01223585334  \
0     0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
1     0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
2     0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
3     0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
4     0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
...   ...  ...     ...           ...   ...   ...          ...          ...   
5567  0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
5568  0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
5569  0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
5570  0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
5571  0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   

      0125698789   02  ...  zhong  zindgi  zoe  zogtorius  zoom  zouk  zyada  \
0            0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
1            0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
2            0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
3            0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
4            0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
...          ...  ...  ...    ...     ...  ...        ...   ...   ...    ...   
5567         0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
5568         0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
5569         0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
5570         0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
5571         0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   

       èn   ú1  〨ud  
0     0.0  0.0  0.0  
1     0.0  0.0  0.0  
2     0.0  0.0  0.0  
3     0.0  0.0  0.0  
4     0.0  0.0  0.0  
...   ...  ...  ...  
5567  0.0  0.0  0.0  
5568  0.0  0.0  0.0  
5569  0.0  0.0  0.0  
5570  0.0  0.0  0.0  
5571  0.0  0.0  0.0  

[5572 rows x 8709 columns]

In [64]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(tfidf_matrix, spam.Category, test_size=0.30, random_state=42)

In [65]:
from sklearn.linear_model import LogisticRegression
lr = LogisticRegression(random_state=42)

In [66]:
lr.fit(X_train, y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=42, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [67]:
y_pred = lr.predict(X_test)

In [69]:
from sklearn.metrics import accuracy_score
accuracy_score(y_test, y_pred)

0.9635167464114832

In [48]:
from sklearn.metrics import confusion_matrix
confusion_matrix(y_test, y_pred)

array([[1446,    2],
       [  59,  165]], dtype=int64)

True positive - 14446 раз модель правильно предсказала, что это не спам, True Negative - 165 раз модель правильно предсказала, что это спам, False Negative - 59 раз ошиблась, был спам, а она сказала, что не спасм, и False Positive - 2 раза ошиблась, был не спам, а можель сказа, что спам
и того получилось, что можель ошиблась 61 раз, что нагдядно показано в табличке ниже

In [47]:
res = pd.DataFrame(y_test, columns= ['Category'])
res['Prediction'] = y_pred
wrong_pred = res[res.Category != res.Prediction]
wrong_pred

Category Prediction
881      spam        ham
1961     spam        ham
3864     spam        ham
2575     spam        ham
3548     spam        ham
...       ...        ...
191      spam        ham
4543     spam        ham
752      spam        ham
495       ham       spam
305      spam        ham

[61 rows x 2 columns]

In [49]:
mes = spam[spam.Message.index.isin(wrong_pred.index)]['Message']

In [45]:
pd.concat([wrong_pred, mes], axis=1)

Category Prediction                                            Message
15       spam        ham  XXXMobileMovieClub: To use your credit, click ...
19       spam        ham  England v Macedonia - dont miss the goals/team...
68       spam        ham  Did you hear about the new "Divorce Barbie"? I...
135      spam        ham  Want 2 get laid tonight? Want real Dogging loc...
139      spam        ham  You'll not rcv any more msgs from the chat svc...
...       ...        ...                                                ...
4652     spam        ham  A £400 XMAS REWARD IS WAITING FOR YOU! Our com...
5037     spam        ham  You won't believe it but it's true. It's Incre...
5120     spam        ham       PRIVATE! Your 2003 Account Statement for 078
5366     spam        ham  A £400 XMAS REWARD IS WAITING FOR YOU! Our com...
5377     spam        ham  The current leading bid is 151. To pause this ...

[61 rows x 3 columns]